rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,379 @@
+we must also support Bucket Rescale rollbacks, even rolling back multiple completed Bucket Rescale operations at once. 
+This requires tracking the instant and corresponding config version for each Bucket Rescale action. 
+During rollback, both the data and config must be reverted to ensure consistency between the configuration and the data state.
+
+So that we need to maintain config versions alongside commit instants which is multiple versions of hashing_config (multi-versioned configs). 
+This necessitates implementing mechanisms for config creation, update, rollback, cleanup and loading
+
+About Hashing_Config Creation
+
+The initial version *must and can only* be created via DDL. During the DDL phase, hoodie will create a 
+`00000000000000000.hashing_config` for the first time.
+
+About Hashing_Config Update
+
+A new version can only be updated via `call partitionBucketIndexManager(overwrite => 'new-expressions', dry-run => 'false')`, which is 
+a replace-commit operation. 
+
+The transactional relationship between the replace-commit operation and hashing_config updates will be explained in the 
+PartitionBucketIndexManager section later.
+
+*Note:* Before updating the hashing_config, users must manually stop all ingestion jobs.
+
+```text
+Firstly DDL phase creates initial version 
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config
+{
+   "rule": "rule-engine",
+   "expressions": "expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+
+Secondly call PartitionBucketIndexManager(overwrite => 'expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2', 'dry-run' = 'false')
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config (deprecated)
+.hoodie/.hashing_meta/20250303095546020.hashing_config (used)
+{
+   "rule": "rule-engine",
+   "expressions": "expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+```
+
+The expression written at the front has a higher priority, and the expression has a higher priority than the default_bucket_number.
+
+About Hashing_Config Rollback
+
+The Bucket Rescale operation is essentially an insert overwrite action and can seamlessly integrate with Hudi's rollback mechanism. 
+One simplification here is that when rolling back a Bucket Rescale, the corresponding replace-commit.hashing_config is not immediately deleted. 
+Instead, its cleanup is deferred to the version cleanup process (discussed later).
+
+This approach achieves two goals:
+1. Minimizes modifications to the rollback service logic.
+2. Lazy cleanup does not impact config loading, as Hudi always loads the latest hashing_config associated with committed replace commit.
+
+About Hashing_Config Cleanup
+
+The cleanup of hashing_config versions (retaining the latest three) will be handled within the Archive Table Service, 
```

Comment:
I feel like we should not bind the cleaning of hashing_config with the archival process. Once the replace_commit is archived, its hashing_config also got removed, while scanning the archive timeline, there is no info to bookeep this config as of now, and we should keep at least the latest version hashing_config.

Reply:
Changed. We will rollback hashing_config during hudi rollback table service.

As for hashing_config archive. Is that make sense we try to merge `out of date hashing_config` into `.hoodie/.hashing_meta/archive/archived.hashing_config` (create tmp and rename) during call bucket rescale  command?

Reply:
> Is that make sense we try to merge out of date hashing_config into

No

Reply:
sync with danny , Version management occurs within the Hashing_config writer and does not rely on the archive service. For implementation of the hashing_config archive, use `.hoodie/.hashing_meta/archive/archived.hashing_config` to retain all historical information. File operations for this are performed by the writer after completion, following a process where data is first written to a temporary file and then renamed.